azext_iot/monitor/property.py

```diff
@@ -7,91 +7,207 @@
+)
+from azext_iot.monitor.parsers.issue import IssueHandler
+
+
+class PropertyMonitor:
+    def __init__(
+        self, cmd, app_id, device_id, central_dns_suffix=CENTRAL_ENDPOINT,
+    ):
+        self._cmd = cmd
+        self._app_id = app_id
+        self._device_id = device_id
+        self._central_dns_suffix = central_dns_suffix
+        self._device_twin_provider = CentralDeviceTwinProvider(
+            cmd=self._cmd, app_id=self._app_id, device_id=self._device_id
+        )
+        self._central_device_provider = CentralDeviceProvider(self._cmd, self._app_id)
+        self._central_template_provider = CentralDeviceTemplateProvider(
+            cmd=self._cmd, app_id=self._app_id
+        )
+        self._template = self._get_device_template()
 
+    def _compare_properties(self, prev_prop: Property, prop: Property):
+        if prev_prop.version == prop.version:
+            return
 
-def start_property_monitor(
-    cmd,
-    device_id,
-    app_id,
-    central_dns_suffix=CENTRAL_ENDPOINT,
-    polling_interval_seconds=DEVICETWIN_POLLING_INTERVAL_SEC,
-):
-    prev_twin = None
+        changes = {
+            key: self._changed_props(prop.props[key], prop.metadata[key], key,)
+            for key, val in prop.metadata.items()
+            if self._is_relevant(key, val)
+        }
 
-    device_twin_provider = CentralDeviceTwinProvider(
-        cmd=cmd, app_id=app_id, device_id=device_id
-    )
+        return changes
 
-    while True:
+    def _is_relevant(self, key, val):
+        if key in {"$lastUpdated", "$lastUpdatedVersion"}:
+            return False
 
-        raw_twin = device_twin_provider.get_device_twin(
-            central_dns_suffix=central_dns_suffix
+        updated_within = datetime.datetime.now() - datetime.timedelta(
+            seconds=DEVICETWIN_MONITOR_TIME_SEC
         )
 
-        twin = DeviceTwin(raw_twin)
-        if prev_twin:
-            change_d = compare_properties(
-                prev_twin.desired_property, twin.desired_property
+        last_updated = isodate.parse_datetime(val["$lastUpdated"])
+        return last_updated.timestamp() >= updated_within.timestamp()
+
+    def _changed_props(self, prop, metadata, property_name):
+        # not an interface - whole thing is change log
+        if not self._is_interface(property_name):
+            return prop
+        # iterate over property in the interface
+        # if the property is not an exact match for what is present in the previous set of properties
+        # track it as a change
+        diff = {
+            key: prop[key]
+            for key, val in metadata.items()
+            if self._is_relevant(key, val)
+        }
+        return diff
+
+    def _validate_payload(self, changes, minimum_severity):
+        for value in changes:
+            issues = self._validate_payload_against_interfaces(
+                changes[value], value, minimum_severity
+            )
+            for issue in issues:
```

Comment:
bug: move `for issue in issues` back one indentation, its currently being done for each `value in changes`

Reply:
Chatted offline, this is intentional. No bug #closed